core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -465,6 +459,20 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
                         logger.debug("connected to node [{}]", node);
                     }
                     transportServiceAdapter.onNodeConnected(node);
```

Comment:
paranoia: if we hit an exception here, we can leave the nodeChannels in the connectedNodes map. Should we move this:

```
  if (connectedNodes.remove(node, nodeChannels)) {   
      transportServiceAdapter.onNodeDisconnected(node);
  }
```

to the finally clause (under the check of success)?